src/machine_learning/loss_function/negative_log_likelihood.rs

```diff
@@ -0,0 +1,77 @@
+// Negative Log Likelihood Loss Function
+//
+// The `neg_log_likelihood` function calculates the Negative Log Likelyhood loss,
+// which is a loss function used for classification problems in machine learning.
+//
+// ## Formula
+//
+// For a pair of actual and predicted values, represented as vectors `y_true` and
+// `y_pred`, the Negative Log Likelihood loss is calculated as:
+//
+// - loss = `-y_true * log(y_pred) - (1 - y_true) * log(1 - y_pred)`.
+//
+// It returns the average loss by dividing the `total_loss` by total no. of
+// elements.
+
+pub fn neg_log_likelihood(y_true: &[f64], y_pred: &[f64]) -> Option<f64> {
+    // Checks if the length of the actual and predicted values are equal
+    if y_true.len() != y_pred.len() || y_pred.is_empty() || y_true.is_empty() {
+        return None;
+    }
+    // Checks values are between 0 and 1
+    for (p, a) in y_pred.iter().zip(y_true.iter()) {
+        if *p < 0.0 || *p > 1.0 || *a < 0.0 || *a > 1.0 {
+            return None;
+        }
+    }
+
+    let mut total_loss: f64 = 0.0;
+    for (p, a) in y_pred.iter().zip(y_true.iter()) {
+        let loss: f64 = -a * p.ln() - (1.0 - a) * (1.0 - p).ln();
+        total_loss += loss;
+    }
+    Some(total_loss / (y_pred.len() as f64))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_neg_log_likelihood() {
+        let actual_values: Vec<f64> = vec![1.0, 0.0, 1.0];
+        let predicted_values: Vec<f64> = vec![0.9, 0.1, 0.8];
+        assert_eq!(
+            neg_log_likelihood(&actual_values, &predicted_values),
+            Some(0.14462152754328741)
+        );
+    }
```

Comment:
Please add some more examples. Use similar approach as suggested below.

Reply:
@jkauerl: and here.

Reply:
Added more examples.

Reply:
> Added more examples.

Where are they? Your last commit, 65671b79add3ba11196bf690df47cfe2c50e7743, is from 19.06.2024.

Reply:
I implemented 7 tests checking for correct calculation, invalid length, invalid values, empty values and negative values. Is there any more tests missing?